test/src/utils/quaternion.test.cpp

```diff
@@ -0,0 +1,353 @@
+}
+
+/**
+ * @brief Test to determine the correct instantiation of a non zero
+ * Quaternion class instance.
+ */
+TEST_F(QuaternionFunctionality, handlesConstructionOfNonZeroQuaternion)
+{
+  int component_array[kSizeOfQuaternion];
+  for (int i = 0; i < kSizeOfQuaternion; i++) {
+    component_array[i] = rand()%1000;
+    while (component_array[i] == 0) {
+      component_array[i] = rand()%1000;
+    }
+  }
+  Quaternion<int> nonZeroQuaternion = Quaternion<int>(
+    component_array[0],
+    component_array[1],
+    component_array[2],
+    component_array[3]);
+  for (int i = 0; i < kSizeOfQuaternion; i++) {
+    ASSERT_EQ(nonZeroQuaternion[i], component_array[i]);
+  }
+}
+
+/**
+ * @brief Test to determine the correct construction of Quaternion
+ * instances by Vector instances. This test includes contruction by
+ * a 4D Vector, 3D Vector and constant, and 3D Vector.
+ */
+TEST_F(QuaternionFunctionality, handlesConstructionByVector)
+{
+  // Testing construction by 4D Vector
+  test_vector = Vector<int, 4>(createRandomArrayForQuaternion());
+  Quaternion<int> quaternion_vector = Quaternion<int>(test_vector);
+  for (int i = 0; i < kSizeOfQuaternion; i++) {
+    ASSERT_EQ(test_vector[i], quaternion_vector[i]);
+  }
+
+  std::array<int, 3> component_array;
+  for (int i = 0; i < 3; i++) {
+    component_array[i] = rand()%1000;
+    while (component_array[i] == 0) {
+      component_array[i] = rand()%1000;
+    }
+  }
+  // Testing  construction by 3D Vector
+  test_vector_3d = Vector<int, 3>(component_array);
+  Quaternion<int> quaternion_vector_3d = Quaternion<int>(test_vector_3d);
+  ASSERT_EQ(0, quaternion_vector_3d[0]);
+  for (int i = 1; i < kSizeOfQuaternion; i++) {
+    ASSERT_EQ(quaternion_vector_3d[i], test_vector_3d[i-1]);
+  }
+
+  // Testing construction by 3D Vector and constant
+  Quaternion<int> quaternion_vector_3d_const = Quaternion<int>(kValue, test_vector_3d);
+  ASSERT_EQ(kValue, quaternion_vector_3d_const[0]);
+  for (int i = 1; i < kSizeOfQuaternion; i++) {
+    ASSERT_EQ(quaternion_vector_3d_const[i], test_vector_3d[i-1]);
+  }
+}
+
+/**
+ * @brief Test to determine whether the Quaternion class supports
+ * addition with a scalar constant.
+ * The test is performed with randomly instantiated quaternions and
+ * random constant value
+ */
+TEST_F(QuaternionFunctionality, handlesAutoAdditionByConstant)
+{
+    quaternion_result_one = quaternion_one + kValue;
+    quaternion_result_two = kValue + quaternion_one;
+    for (int i = 0; i < kSizeOfQuaternion; i++) {
+        ASSERT_EQ(quaternion_result_one[i], quaternion_one[i] + kValue);
+        ASSERT_EQ(quaternion_result_two[i], quaternion_one[i] + kValue);
+    }
+}
+
+/**
+ * @brief Test to determine whether the Quaternion class supports
+ * the incrementation operation on constant values in the normal way
+ * (adding a constant to each component of the Quaternion)
+ */
+TEST_F(QuaternionFunctionality, handlesAdditionByConstant)
+{
+  quaternion_result_one += kValue;
```

Comment:
maybe it is worth creating another one called identity instead of using that name

Reply:
Can you explain this comment more?

Reply:
@pmorande27  Any updates on this?

Reply:
like it is a bit difficult to understand what is quaternion_result_one, so maybe creating another one that is called identity is more clear